lxd/device/device_utils_disk.go

```diff
@@ -258,49 +260,105 @@ func DiskVMVirtfsProxyStart(pidPath string, sharePath string) (func(), *os.File,
 	if err != nil {
 		return nil, nil, fmt.Errorf("Failed to create unix listener for virtfs-proxy-helper: %w", err)
 	}
-	revert.Add(func() { listener.Close() })
+	defer listener.Close()
 
 	cDial, err := net.Dial("unix", listener.Addr().String())
 	if err != nil {
 		return nil, nil, fmt.Errorf("Failed to connect to virtfs-proxy-helper unix listener: %w", err)
 	}
-	revert.Add(func() { cDial.Close() })
+	defer cDial.Close()
 
 	cDialUnix, ok := cDial.(*net.UnixConn)
 	if !ok {
 		return nil, nil, fmt.Errorf("Dialled virtfs-proxy-helper connection isn't unix socket")
 	}
+	defer cDialUnix.Close()
 
 	cDialUnixFile, err := cDialUnix.File()
 	if err != nil {
 		return nil, nil, fmt.Errorf("Failed getting virtfs-proxy-helper unix dialed file: %w", err)
 	}
+	revert.Add(func() { cDialUnixFile.Close() })
 
 	cAccept, err := listener.Accept()
 	if err != nil {
 		return nil, nil, fmt.Errorf("Failed to accept connection to virtfs-proxy-helper unix listener: %w", err)
 	}
-	revert.Add(func() { cAccept.Close() })
-	listener.Close()
+	defer cAccept.Close()
 
 	cAcceptUnix, ok := cAccept.(*net.UnixConn)
 	if !ok {
 		return nil, nil, fmt.Errorf("Accepted virtfs-proxy-helper connection isn't unix socket")
 	}
+	defer cAcceptUnix.Close()
 
 	acceptFile, err := cAcceptUnix.File()
 	if err != nil {
 		return nil, nil, fmt.Errorf("Failed getting virtfs-proxy-helper unix listener file: %w", err)
 	}
+	defer acceptFile.Close()
+
+	var args []string
+	var fdFiles []*os.File
+
+	if len(idmaps) > 0 {
+		rUIDMapPipe, wUIDMapPipe, err := os.Pipe()
+		if err != nil {
+			return nil, nil, err
+		}
+		defer rUIDMapPipe.Close()
+		defer wUIDMapPipe.Close()
+
+		rGIDMapPipe, wGIDMapPipe, err := os.Pipe()
+		if err != nil {
+			return nil, nil, err
+		}
+		defer rGIDMapPipe.Close()
+		defer wGIDMapPipe.Close()
+
+		for _, idmap := range idmaps {
+			if idmap.Isuid {
+				_, err = wUIDMapPipe.WriteString(fmt.Sprintf("%d %d %d\n", idmap.Nsid, idmap.Hostid, idmap.Maprange))
```

Comment:
This needs to be terminated with `\x00` so C can recognize it as a proper C string when reading from the pipe iirc.

Reply:
Nm, I see you're doing it below.

Reply:
Yeah it caught me out first of all, because when passing multiple idmaps, only the first one was being used.